Cannot use Certificates for SSH

Hi,
recently I found out about SSH-Cerificates, so I tried using them in tandem with the trezor-agent. So I created a SSH-CA, exported the public key from my Trezor and signed it. From my perspective everyting went as normal until I confirmed the hidden wallet on my Trezor. As I tried to login using the certificate the following log was generated:

1) Create a sub-shell to use the ssh command as provided by a tutorial
```
user@ubuntu:~$ trezor-agent user@ubuntu2 -s -vv
2021-12-04 12:42:44,164 DEBUG        parsed identity: {'proto': None, 'user': 'user', 'host': 'ubuntu2', 'port': None, 'path': None}      [interface.py:30]
2021-12-04 12:42:44,164 DEBUG        identity parts: ['ssh://', 'user@', 'ubuntu2']                                                       [interface.py:46]
2021-12-04 12:42:44,165 INFO         identity #0: <ssh://user@ubuntu2|nist256p1>                                                          [__init__.py:273]
2021-12-04 12:42:44,172 INFO         using [b'ttyname=/dev/pts/0', b'display=:0'] for pinentry options                                    [ui.py:91]
2021-12-04 12:42:44,602 DEBUG        local SSH version: b'OpenSSH_8.2p1 Ubuntu-4ubuntu0.1, OpenSSL 1.1.1f  31 Mar 2020\n'                 [__init__.py:132]
2021-12-04 12:42:44,603 DEBUG        serving on /tmp/trezor-ssh-agent-2kz9c27z                                                            [server.py:30]
2021-12-04 12:42:44,605 DEBUG        server thread started                                                                                [server.py:121]
2021-12-04 12:42:44,606 INFO         running '/bin/bash' with {'SSH_AUTH_SOCK': '/tmp/trezor-ssh-agent-2kz9c27z', 'SSH_AGENT_PID': '6058'} [server.py:156]
2021-12-04 12:42:44,624 DEBUG        waiting for connection on /tmp/trezor-ssh-agent-2kz9c27z                                             [server.py:129]
2021-12-04 12:42:44,628 DEBUG        subprocess 6063 is running                                                                           [server.py:163]
```

2) Use the tutorial provided ssh-command
```
user@ubuntu:~$ ssh -o CertificateFile=/home/user/Desktop/user-cert.pub user@ubuntu2
2021-12-04 12:42:59,102 DEBUG        welcome agent                                                                                        [server.py:90]
2021-12-04 12:42:59,102 DEBUG        waiting for connection on /tmp/trezor-ssh-agent-2kz9c27z                                             [server.py:129]
2021-12-04 12:42:59,103 DEBUG        request: 1 bytes                                                                                     [protocol.py:94]
2021-12-04 12:42:59,103 DEBUG        calling list_pubs()                                                                                  [protocol.py:102]
2021-12-04 12:42:59,587 INFO         Enumerating HidTransport: found 0 devices                                                            [__init__.py:120]
2021-12-04 12:42:59,593 DEBUG        Starting new HTTP connection (1): 127.0.0.1:21325                                                    [connectionpool.py:221]
2021-12-04 12:42:59,597 DEBUG        http://127.0.0.1:21325 "POST /configure HTTP/1.1" 200 21                                             [connectionpool.py:428]
2021-12-04 12:42:59,612 DEBUG        http://127.0.0.1:21325 "POST /enumerate HTTP/1.1" 200 94                                             [connectionpool.py:428]
2021-12-04 12:42:59,614 INFO         Enumerating BridgeTransport: found 1 devices                                                         [__init__.py:120]
2021-12-04 12:42:59,648 INFO         Enumerating WebUsbTransport: found 1 devices                                                         [__init__.py:120]
2021-12-04 12:42:59,648 INFO         Enumerating UdpTransport: found 0 devices                                                            [__init__.py:120]
2021-12-04 12:42:59,649 DEBUG        using transport: bridge:1                                                                            [trezor.py:53]
2021-12-04 12:42:59,649 INFO         creating client instance for device: bridge:1                                                        [client.py:88]
2021-12-04 12:43:00,628 DEBUG        http://127.0.0.1:21325 "POST /acquire/1/null HTTP/1.1" 200 17                                        [connectionpool.py:428]
2021-12-04 12:43:00,629 DEBUG        sending message: Initialize                                                                          [client.py:116]
2021-12-04 12:43:00,638 DEBUG        http://127.0.0.1:21325 "POST /post/13 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:00,686 DEBUG        http://127.0.0.1:21325 "POST /read/13 HTTP/1.1" 200 424                                              [connectionpool.py:428]
2021-12-04 12:43:00,687 DEBUG        received message: Features                                                                           [client.py:139]
2021-12-04 12:43:00,743 DEBUG        http://127.0.0.1:21325 "POST /release/13 HTTP/1.1" 200 17                                            [connectionpool.py:428]
2021-12-04 12:43:00,744 DEBUG        connected to Trezor CD********************F8                                                         [trezor.py:33]
2021-12-04 12:43:00,744 DEBUG        label    : trezor                                                                                    [trezor.py:34]
2021-12-04 12:43:00,744 DEBUG        vendor   : trezor.io                                                                                 [trezor.py:35]
2021-12-04 12:43:00,744 DEBUG        version  : 1.10.3                                                                                    [trezor.py:39]
2021-12-04 12:43:00,744 DEBUG        revision : b'9276b1702361f70e094286e2f89e919d8a230d5c'                                               [trezor.py:40]
2021-12-04 12:43:01,752 DEBUG        http://127.0.0.1:21325 "POST /acquire/1/null HTTP/1.1" 200 17                                        [connectionpool.py:428]
2021-12-04 12:43:01,753 DEBUG        sending message: GetAddress                                                                          [client.py:116]
2021-12-04 12:43:01,772 DEBUG        http://127.0.0.1:21325 "POST /post/14 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:01,920 DEBUG        http://127.0.0.1:21325 "POST /read/14 HTTP/1.1" 200 12                                               [connectionpool.py:428]
2021-12-04 12:43:01,921 DEBUG        received message: PassphraseRequest                                                                  [client.py:139]
2021-12-04 12:43:02,121 DEBUG        ['pinentry'] -> b'OK Pleased to meet you\n'                                                          [ui.py:109]
2021-12-04 12:43:02,121 DEBUG        ['pinentry'] <- b'SETTITLE Trezor passphrase\n'                                                      [ui.py:97]
2021-12-04 12:43:02,122 DEBUG        ['pinentry'] -> b'OK\n'                                                                              [ui.py:109]
2021-12-04 12:43:02,122 DEBUG        ['pinentry'] <- b'SETPROMPT Passphrase:\n'                                                           [ui.py:97]
2021-12-04 12:43:02,124 DEBUG        ['pinentry'] -> b'OK\n'                                                                              [ui.py:109]
2021-12-04 12:43:02,124 DEBUG        setting 2 options                                                                                    [ui.py:140]
2021-12-04 12:43:02,125 DEBUG        ['pinentry'] <- b'OPTION ttyname=/dev/pts/0\n'                                                       [ui.py:97]
2021-12-04 12:43:02,125 DEBUG        ['pinentry'] -> b'OK\n'                                                                              [ui.py:109]
2021-12-04 12:43:02,125 DEBUG        ['pinentry'] <- b'OPTION display=:0\n'                                                               [ui.py:97]
2021-12-04 12:43:02,125 DEBUG        ['pinentry'] -> b'OK\n'                                                                              [ui.py:109]
2021-12-04 12:43:02,126 DEBUG        ['pinentry'] <- b'GETPIN\n'                                                                          [ui.py:97]
2021-12-04 12:43:04,546 DEBUG        ['pinentry'] -> '**********'                                                                         [ui.py:109]
2021-12-04 12:43:04,551 DEBUG        sending message: PassphraseAck                                                                       [client.py:116]
2021-12-04 12:43:04,598 DEBUG        http://127.0.0.1:21325 "POST /post/14 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:04,904 DEBUG        http://127.0.0.1:21325 "POST /read/14 HTTP/1.1" 200 16                                               [connectionpool.py:428]
2021-12-04 12:43:04,905 DEBUG        received message: ButtonRequest                                                                      [client.py:139]
2021-12-04 12:43:04,905 DEBUG        sending message: ButtonAck                                                                           [client.py:116]
2021-12-04 12:43:04,924 DEBUG        http://127.0.0.1:21325 "POST /post/14 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:07,541 DEBUG        http://127.0.0.1:21325 "POST /read/14 HTTP/1.1" 200 16                                               [connectionpool.py:428]
2021-12-04 12:43:07,544 DEBUG        received message: ButtonRequest                                                                      [client.py:139]
2021-12-04 12:43:07,545 DEBUG        sending message: ButtonAck                                                                           [client.py:116]
2021-12-04 12:43:07,571 DEBUG        http://127.0.0.1:21325 "POST /post/14 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:09,434 DEBUG        http://127.0.0.1:21325 "POST /read/14 HTTP/1.1" 200 84                                               [connectionpool.py:428]
2021-12-04 12:43:09,435 DEBUG        received message: Address                                                                            [client.py:139]
2021-12-04 12:43:09,491 DEBUG        http://127.0.0.1:21325 "POST /release/14 HTTP/1.1" 200 17                                            [connectionpool.py:428]
2021-12-04 12:43:09,492 DEBUG        identity parts: ['ssh://', 'user@', 'ubuntu2']                                                       [interface.py:46]
2021-12-04 12:43:09,494 DEBUG        "<ssh://user@ubuntu2|nist256p1>" getting public key (nist256p1) from Trezor                          [trezor.py:82]
2021-12-04 12:43:09,507 DEBUG        identity parts: ['ssh://', 'user@', 'ubuntu2']                                                       [interface.py:46]
2021-12-04 12:43:09,508 DEBUG        bip32 address string: b'\x00\x00\x00\x00ssh://user@ubuntu2'                                          [interface.py:88]
2021-12-04 12:43:10,485 DEBUG        http://127.0.0.1:21325 "POST /acquire/1/null HTTP/1.1" 200 17                                        [connectionpool.py:428]
2021-12-04 12:43:10,485 DEBUG        sending message: GetPublicKey                                                                        [client.py:116]
2021-12-04 12:43:10,494 DEBUG        http://127.0.0.1:21325 "POST /post/15 HTTP/1.1" 200 0                                                [connectionpool.py:428]
2021-12-04 12:43:10,664 DEBUG        http://127.0.0.1:21325 "POST /read/15 HTTP/1.1" 200 420                                              [connectionpool.py:428]
2021-12-04 12:43:10,665 DEBUG        received message: PublicKey                                                                          [client.py:139]
2021-12-04 12:43:10,722 DEBUG        http://127.0.0.1:21325 "POST /release/15 HTTP/1.1" 200 17                                            [connectionpool.py:428]
2021-12-04 12:43:10,723 DEBUG        result: <PublicKey: {'node': <HDNodeType: {'depth': 5, 'fingerprint': 49*****76, 'child_num': 36******23, 'chain_code': b'}{\x92\x************************************************************************************99re\xbf', 'public_key': b'\x03\x*********************************************************1b)#-\xb0>+N'}>, 'xpub': 'xpub6Fd*******************************************+*********************************************************Nsy'}> [trezor.py:89]
2021-12-04 12:43:10,725 DEBUG        identity parts: ['ssh://', 'user@', 'ubuntu2']                                                       [interface.py:46]
2021-12-04 12:43:10,727 DEBUG        fingerprint: d5:20:***********************************:53:61                                         [formats.py:189]
2021-12-04 12:43:10,727 DEBUG        loading SSH public key: 'ecdsa-sha2-nistp256 AAAAE2V*********************************************************************************************************************************iWc= <ssh://user@ubuntu2|nist256p1>\n' [formats.py:196]
2021-12-04 12:43:10,728 DEBUG        key type: b'ecdsa-sha2-nistp256'                                                                     [formats.py:55]
2021-12-04 12:43:10,728 DEBUG        curve name: b'nistp256'                                                                              [formats.py:62]
2021-12-04 12:43:10,728 DEBUG        loaded ecdsa-sha2-nistp256 public key: d5:20:***********************************:53:61               [formats.py:202]
2021-12-04 12:43:10,729 DEBUG        available keys: [b'<ssh://user@ubuntu2|nist256p1>']                                                  [protocol.py:114]
2021-12-04 12:43:10,729 DEBUG         1) d5:20:***********************************:53:61                                                  [protocol.py:116]
2021-12-04 12:43:10,729 DEBUG        reply: 154 bytes                                                                                     [protocol.py:105]
2021-12-04 12:43:10,921 DEBUG        request: 738 bytes                                                                                   [protocol.py:94]
2021-12-04 12:43:10,922 DEBUG        calling sign_message()                                                                               [protocol.py:102]
2021-12-04 12:43:10,922 DEBUG        key type: b'ecdsa-sha2-nistp256'                                                                     [formats.py:55]
2021-12-04 12:43:10,922 DEBUG        curve name: b'nistp256'                                                                              [formats.py:62]
2021-12-04 12:43:10,923 DEBUG        looking for d5:20:***********************************:53:61                                          [protocol.py:128]
2021-12-04 12:43:10,923 DEBUG        loading SSH public key: 'ecdsa-sha2-nistp256 AAAAE2V*********************************************************************************************************************************iWc= <ssh://user@ubuntu2|nist256p1>\n' [formats.py:196]
2021-12-04 12:43:10,924 DEBUG        key type: b'ecdsa-sha2-nistp256'                                                                     [formats.py:55]
2021-12-04 12:43:10,924 DEBUG        curve name: b'nistp256'                                                                              [formats.py:62]
2021-12-04 12:43:10,925 DEBUG        loaded ecdsa-sha2-nistp256 public key: d5:20:***********************************:53:61               [formats.py:202]
2021-12-04 12:43:10,925 DEBUG        using key b'<ssh://user@ubuntu2|nist256p1>' (d5:20:***********************************:53:61)        [protocol.py:135]
2021-12-04 12:43:10,926 DEBUG        signing 621-byte blob with "<ssh://user@ubuntu2|nist256p1>" key                                      [protocol.py:142]
2021-12-04 12:43:10,927 DEBUG        key type: b'ecdsa-sha2-nistp256-cert-v01@openssh.com'                                                [formats.py:55]
sign_and_send_pubkey: signing failed for ECDSA "<ssh://user@ubuntu2|nist256p1>": communication with agent failed
2021-12-04 12:43:10,928 WARNING      error: b'ecdsa-sha2-nistp256-cert-v01@openssh.com'                                                   [server.py:100]
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.8/site-packages/libagent/server.py", line 95, in handle_connection
    reply = handler.handle(msg=msg)
  File "/home/user/.local/lib/python3.8/site-packages/libagent/ssh/protocol.py", line 103, in handle
    reply = method(buf=buf)
  File "/home/user/.local/lib/python3.8/site-packages/libagent/ssh/protocol.py", line 144, in sign_message
    signature = self.conn.sign(blob=blob, identity=key['identity'])
  File "/home/user/.local/lib/python3.8/site-packages/libagent/ssh/__init__.py", line 234, in sign
    return conn.sign_ssh_challenge(blob=blob, identity=identity)
  File "/home/user/.local/lib/python3.8/site-packages/libagent/ssh/client.py", line 36, in sign_ssh_challenge
    msg = _parse_ssh_blob(blob)
  File "/home/user/.local/lib/python3.8/site-packages/libagent/ssh/client.py", line 63, in _parse_ssh_blob
    res['public_key'] = formats.parse_pubkey(public_key)
  File "/home/user/.local/lib/python3.8/site-packages/libagent/formats.py", line 56, in parse_pubkey
    assert key_type in SUPPORTED_KEY_TYPES, key_type
AssertionError: b'ecdsa-sha2-nistp256-cert-v01@openssh.com'
user@ubuntu2's password: ^C
```

3) Exit from the sub-shell
```
user@ubuntu:~$ exit
exit
2021-12-04 12:44:46,735 DEBUG        subprocess 6063 exited: 130                                                                          [server.py:165]
2021-12-04 12:44:46,737 DEBUG        closing server                                                                                       [__init__.py:148]
2021-12-04 12:44:46,777 DEBUG        server stopped                                                                                       [server.py:133]
2021-12-04 12:44:46,777 DEBUG        server thread stopped                                                                                [server.py:138]

user@ubuntu:~$
```

In this log I blured some informations depending on my Trezor with * and replaced the names of the user and the machines. I hope that this does not matters. If it does I can still post the original log.

--- a/libagent/formats.py
+++ b/libagent/formats.py
@@ -25,8 +25,10 @@ SSH_NIST256_DER_OCTET = b'\x04'
 SSH_NIST256_KEY_PREFIX = b'ecdsa-sha2-'
 SSH_NIST256_CURVE_NAME = b'nistp256'
 SSH_NIST256_KEY_TYPE = SSH_NIST256_KEY_PREFIX + SSH_NIST256_CURVE_NAME
+SSH_NIST256_CERT_POSTFIX = b'-cert-v01@openssh.com'
+SSH_NIST256_CERT_TYPE = SSH_NIST256_KEY_TYPE + SSH_NIST256_CERT_POSTFIX
 SSH_ED25519_KEY_TYPE = b'ssh-ed25519'
-SUPPORTED_KEY_TYPES = {SSH_NIST256_KEY_TYPE, SSH_ED25519_KEY_TYPE}
+SUPPORTED_KEY_TYPES = {SSH_NIST256_KEY_TYPE, SSH_NIST256_CERT_TYPE, SSH_ED25519_KEY_TYPE}
 
 hashfunc = hashlib.sha256
 
@@ -57,10 +59,27 @@ def parse_pubkey(blob):
 
     result = {'blob': blob, 'type': key_type, 'fingerprint': fp}
 
-    if key_type == SSH_NIST256_KEY_TYPE:
+    if key_type == SSH_NIST256_KEY_TYPE or key_type == SSH_NIST256_CERT_TYPE:
+        if key_type == SSH_NIST256_CERT_TYPE:
+            _nonce = util.read_frame(s)
+
         curve_name = util.read_frame(s)
         log.debug('curve name: %s', curve_name)
         point = util.read_frame(s)
+
+        if key_type == SSH_NIST256_CERT_TYPE:
+            _serial_number = util.recv(s, '>Q')
+            _type = util.recv(s, '>L')
+            _key_id = util.read_frame(s)
+            _valid_principals = util.read_frame(s)
+            _valid_after = util.recv(s, '>Q')
+            _valid_before = util.recv(s, '>Q')
+            _critical_options = util.read_frame(s)
+            _extensions = util.read_frame(s)
+            _reserved = util.read_frame(s)
+            _signature_key = util.read_frame(s)
+            _signature = util.read_frame(s)
+
         assert s.read() == b''
         _type, point = point[:1], point[1:]
         assert _type == SSH_NIST256_DER_OCTET

--- a/libagent/tests/test_formats.py
+++ b/libagent/tests/test_formats.py
@@ -23,6 +23,26 @@ _public_key = (
     'home\n'
 )
 
+_public_key_cert = (
+	'ecdsa-sha2-nistp256-cert-v01@openssh.com '
+	'AAAAKGVjZHNhLXNoYTItbmlzdHAyNTYtY2VydC12MDFAb3B'
+	'lbnNzaC5jb20AAAAgohlAP8H3LPYWz3+w/E+RGDxG6tNAEE'
+	'3Ao9Z6Pc66khEAAAAIbmlzdHAyNTYAAABBBGI2zqveJSB+g'
+	'eQEWG46OvGs2h3+0qu7tIdsH8WylrV19vttd7GR5rKvTWJt'
+	'8b9ErthmnFALelAFKOB/u50jsukAAAAAAAAAFQAAAAEAAAA'
+	'IdW5pdFRlc3QAAAAIAAAABHVzZXIAAAAAAAAAAP////////'
+	'//AAAAAAAAAIIAAAAVcGVybWl0LVgxMS1mb3J3YXJkaW5nA'
+	'AAAAAAAABdwZXJtaXQtYWdlbnQtZm9yd2FyZGluZwAAAAAA'
+	'AAAWcGVybWl0LXBvcnQtZm9yd2FyZGluZwAAAAAAAAAKcGV'
+	'ybWl0LXB0eQAAAAAAAAAOcGVybWl0LXVzZXItcmMAAAAAAA'
+	'AAAAAAADMAAAALc3NoLWVkMjU1MTkAAAAgf9gyPrF24CLZc'
+	'0rHoZuI1+yjBFWt66G8oUmm20yRO8IAAABTAAAAC3NzaC1l'
+	'ZDI1NTE5AAAAQCEgVgsR7fSgcTxuAWqMW4h42y7pt1BAKR4'
+	'HTRg178tl7Vx8WoRtQcNirX9eggBcTA+5ILWmeY3uDN+soW'
+	't7fwk= '
+	'home\n'
+)
+
 
 def test_parse_public_key():
     key = formats.import_public_key(_public_key)
@@ -32,6 +52,15 @@ def test_parse_public_key():
     assert key['curve'] == 'nist256p1'
     assert key['fingerprint'] == '4b:19:bc:0f:c8:7e:dc:fa:1a:e3:c2:ff:6f:e0:80:a2'  # nopep8
     assert key['type'] == b'ecdsa-sha2-nistp256'
+
+def test_parse_public_key_cert()
+    key = formats.import_public_key(_public_key_cert)
+    assert key['name'] == b'home'
+    assert key['point'] == _point
+
+    assert key['curve'] == 'nist256p1'
+    assert key['fingerprint'] == '4b:19:bc:0f:c8:7e:dc:fa:1a:e3:c2:ff:6f:e0:80:a2'  # nopep8
+    assert key['type'] == b'ecdsa-sha2-nistp256-cert-v01@openssh.com'
 
 
 def test_decompress():

